src/braket/circuits/circuit.py

```diff
@@ -438,6 +440,136 @@ def add_circuit(
 
         return self
 
+    def add_noise(
+        self,
+        noise: Noise,
+        target_gates: Union[str, Iterable[str]] = None,
+        target_qubits: QubitSetInput = None,
+        target_times: Union[int, Iterable[int]] = None,
+        insert_strategy: str = "inclusive",
+    ) -> Circuit:
+        """
+        Add the provided `noise` to gates, qubits and time specified by `target_gates`,
+         `target_qubits` and `target_times`. When `target_gates`=None, `noise` is added
+        to the specified qubits and time. When `target_qubits`=None, `noise` is added to
+        the specified gates and time at all qubits. When `target_times`=None, `noise` is
+        added to the specified gates and qubits at all time. If all `target_gates`,
+        `target_qubits` and `target_times` are None, `noise` is added to every qubit at
+        every time in a circuit.
+
+        When `target_gates` is not None and `noise.qubit_count`>1, `noise.qubit_count`
+        must be the same as `qubit_count` of gates specified by `target_gates`. When
+        `noise.qubit_count`==1, ie. `noise` is a single-qubit noise, `noise` is added
+        to all qubits in `target_qubits` that interact with the target gates.
+
+        Args:
+            noise (Noise): Noise to be added to the circuit. When `noise.qubit_count`>1,
+                `noise.qubit_count` must be the same as `qubit_count` of gates specified by
+                `target_gates`.
+            targer_gates (Union[str, Iterable[str]): Name or List of name of gates which
+                `noise` is added to. If None, `noise` is added only according to
+                `target_qubits` and `target_times`. None should be used when users want
+                to add `noise` to a ciruit moment that has no gate.
+            target_qubits (QubitSetInput): Index or indices of qubit(s). When `target_gates`
+                is not None, the usage of `target_qubits` is determined by `insert_strategy`.
+                Default=None.
+            target_times (Union[int, Iterable[int]]): Index of indices of time which `noise`
+                is added to. Default=None.
+            insert_strategy (str): Rule of how `target_qubit` is used. `insert_strategy`
+                is usded only when `target_gates` is not None. Default="inclusive".
+                Options:
+                    "strict": Insert noise to a gate when `gate.target` exactly matches
+                        `target_qubits`. Sensitive to the order of qubits.
+                    "inclusive": Insert noise to a gate when `gate.target` is a subset
+                        of `target_qubits`.
+
+        Returns:
+            Circuit: self
+
+        Raises:
```

Comment:
What about when qubit count of noise doesn't match target_gates?

Reply:
The input is the name of the gates (str type). It is not straightforward to check the qubit_count through their name. An idea I have is using `gate = getattr(Gate, name)` and then check the qubit count. Does this looks okay?

Reply:
As Kshitij suggested above, it's easier to pass in the gate and noise objects directly instead of using strings.

Reply:
Follow-up question: We could allow user pass a gate class (e.g. Gate.Rx) or a gate object (e.g. Gate.Rx(angle=0.1)). In the former option, we can't actually access `qubit_count` from Gate.X. In the later option, do you think user would expect all properties of gate have to match or just the name?

Reply:
As a user, if the gate object is passed in, I'd expect that all properties would match, not just the class name.

Reply:
I think a more common use case would be that the user wants to match all gates with a given name (e.g. all XY gates). Do you have a suggestion for a clean way to do this (and also validate qubit_count)?

Reply:
Just has a discussion with Cedric. 
(1) As Cedric pointed above, usually users only want to compare the name of gate. So in this case, passing an instance of Gate seems inappropriate. 
(2) The alternative is to pass class itself without instantiation. In this case, is there a clean way to do type checking? I found that `type(Gate.X)` returns  `<class 'abc.ABCMeta'>` , and `Type[Gate.X]` returns `typing.Type[braket.circuits.gates.X]`. Both of them are not good for type checking. 
(3) Another alternative is just use `str` type as input and user just specify the name of the gates. 

Note that in (2) and (3), we can only check the name of the gate, and have no access to qubit_count.

Reply:
As a follow-up, I will use target_gates.__name__ for comparing the target, and issubclass(target_gates, Gate) for type checking.

Reply:
@speller26 Do you know the standard format for the type hint of a class (not an instance)? For example, how do we make it clear that the input should be Gate.X, not Gate.X()?

Reply:
I think I didn't understand a different question you asked earlier and accidentally answered this one instead. Check out https://pydantic-docs.helpmanual.io/usage/types/#type
The type hint is probably Type[Gate].

Reply:
okay, thanks! I saw this too but I wasn't sure this is a standard thing or just for pydantic tools.